configure fails with --enable-linux-builtin when CONFIG_MODULES=n

### System information
Type | Version/Name
 --- | ---
Distribution Name	| Arch
Distribution Version	| rolling (up to date as of the submission of this bug report)
Linux Kernel	| 5.8.4
Architecture	| x86-64
ZFS Version	| 2.0.0-rc1
SPL Version	| 2.0.0-rc1

### Describe the problem you're observing
After copying my `.config` & running `make prepare` under a copy of the Linux source, then running `autogen.sh` & `configure` with `--enable-linux-builtin` in a clean zfs source tree, I get the error
```
configure: error: 
        *** Unable to build an empty module.
        *** Please run 'make scripts' inside the kernel source tree.
```
I also tried to run `make oldconfig` & `make scripts` in the kernel directory before re-cloning the zfs tree & retrying the above steps, though I do realize that `make prepare` builds those targets automatically - still the same result. Note that I was able to successfully build zfs into the kernel with 5.6/0.8.x just a few months ago, & each time `configure` ran normally.
### Describe how to reproduce the problem
Run `make prepare` in a configured Linux source tree (in this case under /tmp/linux/), then run `sh autogen.sh` & `sh configure --with-config=kernel --with-linux=/tmp/linux/ --with-linux-obj=/tmp/linux/ --enable-linux-builtin`. `configure` will fail.
### Include any warning/errors/backtraces from the system logs
`config.log` (before patch):
```
configure:13570: checking kernel file name for module symbols     
configure:13600: result: NONE                                                  
configure:13701: checking whether modules can be built             
configure:13762:                     
            KBUILD_MODPOST_NOFINAL= KBUILD_MODPOST_WARN=     
            make modules -k -j16 -C /tmp/linux/                         
            M=/tmp/zfs/build/conftest >build/conftest/build.log 2>&1     
configure:13765: $? = 2                                                   
configure:13768: test -f build/conftest/conftest.o                  
configure:13771: $? = 1     
configure:13779: result: no     
configure:13786: error:      
        *** Unable to build an empty module.     
        *** Please run 'make scripts' inside the kernel source tree.
```
`build/conftest/build.log`:
```
make: Entering directory '/tmp/linux'    
make[1]: *** No rule to make target '/tmp/zfs/build/conftest/modules.order', needed by '/tmp/zfs/build/conftest/Module.symvers'.    
WARNING: Symbol version dump "Module.symvers" is missing.    
         Modules may not have dependencies or modversions.    
make[1]: Target '__modpost' not remade because of errors.    
make: *** [Makefile:1672: modules] Error 2    
make: Leaving directory '/tmp/linux'
```

I can confirm this error with ZFS version 2.0.5 and kernel-5.10.62.

config.log:
```
configure:17538: checking kernel source directory
configure:17581: result: /usr/src/linux
configure:17592: checking kernel build directory
configure:17619: result: /usr/src/linux
configure:17622: checking kernel source version
configure:17670: result: 5.10.62-zfs
configure:17702: checking kernel file name for module symbols
configure:17732: result: NONE
configure:17833: checking whether modules can be built
configure:17901:
            KBUILD_MODPOST_NOFINAL= KBUILD_MODPOST_WARN=
            make modules -k -j4 -C /usr/src/linux
            M=/var/tmp/portage/sys-fs/zfs-kmod-2.0.5/work/zfs-2.0.5/build/conftest >build/conftest/buil
d.log 2>&1
configure:17904: $? = 2
configure:17907: test -f build/conftest/conftest.o
configure:17910: $? = 1
configure:17918: result: no
configure:17925: error:
        *** Unable to build an empty module.
        *** Please run 'make scripts' inside the kernel source tree.
```

build/conftest/build.log:
```
make: Entering directory '/usr/src/linux-5.10.62'
make[1]: *** No rule to make target '/var/tmp/portage/sys-fs/zfs-kmod-2.0.5/work/zfs-2.0.5/build/conftest/modules.order', needed by '/var/tmp/portage/sys-fs/zfs-kmod-2.0.5/work/zfs-2.0.5/build/conftest/Module.symvers'.
WARNING: Symbol version dump "Module.symvers" is missing.
         Modules may not have dependencies or modversions.
make[1]: Target '__modpost' not remade because of errors.
make: *** [Makefile:1726: modules] Error 2
make: Leaving directory '/usr/src/linux-5.10.62'
```

After modifying "conftest" not to fail, I get additional failures of configuration tests:
```
checking whether mutex_lock() is GPL-only... yes
configure: error:
	*** Kernel built with CONFIG_DEBUG_LOCK_ALLOC which is incompatible
	*** with the CDDL license and will prevent the module linking stage
	*** from succeeding.  You must rebuild your kernel without this
	*** option enabled.

```
... which is wrong:
```
# grep CONFIG_DEBUG_LOCK_ALLOC /usr/src/linux/.config
# CONFIG_DEBUG_LOCK_ALLOC is not set
```
Again modifying this test not to fail:

```
configure: error:
	*** This kernel does not include the required zlib inflate support.
	*** Rebuild the kernel with CONFIG_ZLIB_INFLATE=y|m set.
```
... which is again wrong:
```
# grep CONFIG_ZLIB_INFLATE /usr/src/linux/.config
CONFIG_ZLIB_INFLATE=y
```

RoelofSol's change of ZFS_LINUX_COMPILE does not fix this.

I see there was talk in previous similar bug reports of adding CONFIG_MODULES=n to the integration testing. Is there some progress on that?

--- a/config/kernel-config-defined.m4
+++ b/config/kernel-config-defined.m4
@@ -19,6 +19,7 @@ AC_DEFUN([ZFS_AC_KERNEL_CONFIG_DEFINED], [
 		])
 	])
 
+	ZFS_AC_KERNEL_SRC_CONFIG_MODULES
 	ZFS_AC_KERNEL_SRC_CONFIG_BLOCK
 	ZFS_AC_KERNEL_SRC_CONFIG_DEBUG_LOCK_ALLOC
 	ZFS_AC_KERNEL_SRC_CONFIG_TRIM_UNUSED_KSYMS
@@ -29,6 +30,7 @@ AC_DEFUN([ZFS_AC_KERNEL_CONFIG_DEFINED], [
 	ZFS_LINUX_TEST_COMPILE_ALL([config])
 	AC_MSG_RESULT([done])
 
+	ZFS_AC_KERNEL_CONFIG_MODULES
 	ZFS_AC_KERNEL_CONFIG_BLOCK
 	ZFS_AC_KERNEL_CONFIG_DEBUG_LOCK_ALLOC
 	ZFS_AC_KERNEL_CONFIG_TRIM_UNUSED_KSYMS
@@ -96,6 +98,61 @@ AC_DEFUN([ZFS_AC_KERNEL_CONFIG_DEBUG_LOCK_ALLOC], [
 	*** with the CDDL license and will prevent the module linking stage
 	*** from succeeding.  You must rebuild your kernel without this
 	*** option enabled.])
+	])
+])
+
+dnl #
+dnl # Check CONFIG_MODULES
+dnl #
+dnl # Verify the kernel has CONFIG_MODULES support enabled.
+dnl #
+AC_DEFUN([ZFS_AC_KERNEL_SRC_CONFIG_MODULES], [
+	ZFS_LINUX_TEST_SRC([config_modules], [
+		#if !defined(CONFIG_MODULES)
+		#error CONFIG_MODULES not defined
+		#endif
+	],[])
+])
+
+AC_DEFUN([ZFS_AC_KERNEL_CONFIG_MODULES], [
+	AC_MSG_CHECKING([whether CONFIG_MODULES is defined])
+	AS_IF([test "x$enable_linux_builtin" != xyes], [
+		ZFS_LINUX_TEST_RESULT([config_modules], [
+			AC_MSG_RESULT([yes])
+		],[
+			AC_MSG_RESULT([no])
+			AC_MSG_ERROR([
+		*** This kernel does not include the required loadable module
+		*** support!
+		***
+		*** To build OpenZFS as a loadable Linux kernel module
+		*** enable loadable module support by setting
+		*** `CONFIG_MODULES=y` in the kernel configuration and run
+		*** `make modules_prepare` in the Linux source tree.
+		***
+		*** If you don't intend to enable loadable kernel module
+		*** support, please compile OpenZFS as a Linux kernel built-in.
+		***
+		*** Prepare the Linux source tree by running `make prepare`,
+		*** use the OpenZFS `--enable-linux-builtin` configure option,
+		*** copy the OpenZFS sources into the Linux source tree using
+		*** `./copy-builtin <linux source directory>`,
+		*** set `CONFIG_ZFS=y` in the kernel configuration and compile
+		*** kernel as usual.
+			])
+		])
+	], [
+		ZFS_LINUX_TRY_COMPILE([], [], [
+			AC_MSG_RESULT([not needed])
+		],[
+			AC_MSG_RESULT([error])
+			AC_MSG_ERROR([
+		*** This kernel is unable to compile object files.
+		***
+		*** Please make sure you prepared the Linux source tree
+		*** by running `make prepare` there.
+			])
+		])
 	])
 ])
 

--- a/config/kernel.m4
+++ b/config/kernel.m4
@@ -8,8 +8,8 @@ AC_DEFUN([ZFS_AC_CONFIG_KERNEL], [
 		ZFS_AC_QAT
 
 		dnl # Sanity checks for module building and CONFIG_* defines
-		ZFS_AC_KERNEL_TEST_MODULE
 		ZFS_AC_KERNEL_CONFIG_DEFINED
+		ZFS_AC_MODULE_SYMVERS
 
 		dnl # Sequential ZFS_LINUX_TRY_COMPILE tests
 		ZFS_AC_KERNEL_FPU_HEADER
@@ -443,8 +443,6 @@ AC_DEFUN([ZFS_AC_KERNEL], [
 	AC_SUBST(LINUX)
 	AC_SUBST(LINUX_OBJ)
 	AC_SUBST(LINUX_VERSION)
-
-	ZFS_AC_MODULE_SYMVERS
 ])
 
 dnl #
@@ -536,27 +534,6 @@ AC_DEFUN([ZFS_AC_QAT], [
 	$QAT_SYMBOLS
 			])
 		])
-	])
-])
-
-dnl #
-dnl # Basic toolchain sanity check.
-dnl #
-AC_DEFUN([ZFS_AC_KERNEL_TEST_MODULE], [
-	AC_MSG_CHECKING([whether modules can be built])
-	ZFS_LINUX_TRY_COMPILE([], [], [
-		AC_MSG_RESULT([yes])
-	],[
-		AC_MSG_RESULT([no])
-		if test "x$enable_linux_builtin" != xyes; then
-			AC_MSG_ERROR([
-	*** Unable to build an empty module.
-			])
-		else
-			AC_MSG_ERROR([
-	*** Unable to build an empty module.
-	*** Please run 'make scripts' inside the kernel source tree.])
-		fi
 	])
 ])
 
@@ -662,8 +639,10 @@ AC_DEFUN([ZFS_LINUX_COMPILE], [
 		build kernel modules with LLVM/CLANG toolchain])
 	AC_TRY_COMMAND([
 	    KBUILD_MODPOST_NOFINAL="$5" KBUILD_MODPOST_WARN="$6"
-	    make modules -k -j$TEST_JOBS ${KERNEL_CC:+CC=$KERNEL_CC} ${KERNEL_LD:+LD=$KERNEL_LD} ${KERNEL_LLVM:+LLVM=$KERNEL_LLVM} -C $LINUX_OBJ $ARCH_UM
-	    M=$PWD/$1 >$1/build.log 2>&1])
+	    make modules -k -j$TEST_JOBS ${KERNEL_CC:+CC=$KERNEL_CC}
+	    ${KERNEL_LD:+LD=$KERNEL_LD} ${KERNEL_LLVM:+LLVM=$KERNEL_LLVM}
+	    CONFIG_MODULES=y CFLAGS_MODULE=-DCONFIG_MODULES
+	    -C $LINUX_OBJ $ARCH_UM M=$PWD/$1 >$1/build.log 2>&1])
 	AS_IF([AC_TRY_COMMAND([$2])], [$3], [$4])
 ])
 

